File: python/cli/sekreto_cli.py
#!/usr/bin/env python3
# A tiny app that needs a secret.
#
# It asks sekreto for `api.token` and calls the token-protected API with
# it. Every port ships this same CLI, and test/integration.sh runs all of
# them against the same server from all four secret sources - which is
# what proves the library, rather than the spec alone.
#
# Usage: sekreto_cli.py <api-url> [--source env|dotenv|hashicorp|boru|chain]
#                                 [--store <name>]   directed read

import json
import os
import sys
import urllib.error
import urllib.request

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))

from voxgig_sekreto import Sekreto  # noqa: E402

LANG = 'python'


def chainfor(source):
    env = os.environ

    envspec = {'kind': 'env', 'prefix': env.get('SEKRETO_PREFIX')}
    dotenvspec = {'kind': 'dotenv', 'file': env.get('SEKRETO_DOTENV') or '.env'}
    hashicorpspec = {
        'kind': 'hashicorp',
        'addr': env.get('VAULT_ADDR') or '',
        'token': env.get('VAULT_TOKEN') or '',
        'mount': env.get('VAULT_MOUNT'),
    }
    boruspec = {
        'kind': 'boru',
        'command': env.get('BORU_COMMAND') or 'boru',
        'namespace': env.get('BORU_NAMESPACE'),
        'home': env.get('BORU_HOME'),
    }

    if 'env' == source:
        return [envspec]
    if 'dotenv' == source:
        return [dotenvspec]
    if 'hashicorp' == source:
        return [hashicorpspec]
    if 'boru' == source:
        return [boruspec]

    # The default: the chain an app would actually ship with - local
    # overrides first, shared vaults last.
    return [envspec, dotenvspec, hashicorpspec, boruspec]


def main():
    args = sys.argv[1:]
    url = args[0] if args else 'http://127.0.0.1:8099/whoami'

    source = args[args.index('--source') + 1] if '--source' in args else 'chain'

    # --store names a store outright: the secret must come from that one,
    # not from whichever provider happens to answer first.
    store = args[args.index('--store') + 1] if '--store' in args else ''

    secrets = Sekreto({'providers': chainfor(source)})

    try:
        token = secrets.getfrom(store, 'api.token') if store else secrets.get('api.token')
    except Exception as err:
        print('sekreto-cli: ' + str(err), file=sys.stderr)
        return 2

    request = urllib.request.Request(
        url,
        headers={'Authorization': 'Bearer ' + token, 'X-Sekreto-Lang': LANG},
        method='GET',
    )

    try:
        with urllib.request.urlopen(request) as response:
            body = json.loads(response.read().decode('utf8'))
    except urllib.error.HTTPError as err:
        # Never print the token itself, even when the call fails.
        print('sekreto-cli: ' + secrets.redact(err.read().decode('utf8')), file=sys.stderr)
        return 1

    print(
        json.dumps(
            {
                'ok': True,
                'lang': LANG,
                'source': source,
                'store': store,
                'caller': body.get('caller'),
            },
            separators=(',', ':'),
        )
    )

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: ruby/cli/sekreto_cli.rb
# frozen_string_literal: true

# A tiny app that needs a secret.
#
# It asks sekreto for `api.token` and calls the token-protected API with
# it. Every port ships this same CLI, and test/integration.sh runs all of
# them against the same server from all four secret sources - which is
# what proves the library, rather than the spec alone.
#
# Usage: ruby sekreto_cli.rb <api-url> [--source env|dotenv|hashicorp|boru|chain]
#                                      [--store <name>]   directed read

require 'json'
require 'net/http'
require 'uri'

require_relative '../lib/voxgig_sekreto'

LANG = 'ruby'

def chainfor(source)
  envspec = { 'kind' => 'env', 'prefix' => ENV.fetch('SEKRETO_PREFIX', nil) }
  dotenvspec = { 'kind' => 'dotenv', 'file' => ENV['SEKRETO_DOTENV'] || '.env' }
  hashicorpspec = {
    'kind' => 'hashicorp',
    'addr' => ENV['VAULT_ADDR'] || '',
    'token' => ENV['VAULT_TOKEN'] || '',
    'mount' => ENV.fetch('VAULT_MOUNT', nil)
  }
  boruspec = {
    'kind' => 'boru',
    'command' => ENV['BORU_COMMAND'] || 'boru',
    'namespace' => ENV.fetch('BORU_NAMESPACE', nil),
    'home' => ENV.fetch('BORU_HOME', nil)
  }

  case source
  when 'env' then [envspec]
  when 'dotenv' then [dotenvspec]
  when 'hashicorp' then [hashicorpspec]
  when 'boru' then [boruspec]
  else
    # The default: the chain an app would actually ship with - local
    # overrides first, shared vaults last.
    [envspec, dotenvspec, hashicorpspec, boruspec]
  end
end

def main
  args = ARGV
  url = args[0] || 'http://127.0.0.1:8099/whoami'

  flag = args.index('--source')
  source = flag.nil? ? 'chain' : args[flag + 1]

  # --store names a store outright: the secret must come from that one, not
  # from whichever provider happens to answer first.
  storeflag = args.index('--store')
  store = storeflag.nil? ? '' : args[storeflag + 1]

  secrets = VoxgigSekreto::Sekreto.new('providers' => chainfor(source))

  begin
    token = store.empty? ? secrets.get('api.token') : secrets.getfrom(store, 'api.token')
  rescue StandardError => e
    warn 'sekreto-cli: ' + e.message
    return 2
  end

  uri = URI.parse(url)
  request = Net::HTTP::Get.new(uri)
  request['Authorization'] = 'Bearer ' + token
  request['X-Sekreto-Lang'] = LANG

  response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: 'https' == uri.scheme) do |http|
    http.request(request)
  end

  if '200' != response.code
    # Never print the token itself, even when the call fails.
    warn 'sekreto-cli: ' + secrets.redact(response.body.to_s)
    return 1
  end

  body = JSON.parse(response.body)

  puts JSON.generate({ 'ok' => true, 'lang' => LANG, 'source' => source,
                       'store' => store, 'caller' => body['caller'] })

  0
end

exit(main)
